linkerd/app/inbound/src/direct.rs

```diff
@@ -98,31 +102,58 @@ impl<N> Inbound<N> {
                 // target name as a part of the header.
-                //
-                // TODO: Apply port policies. This isn't necessary for now, since these connections
-                // always have a client identity. We'll need to honor client restrictions once those
-                // are supported, though.
                 .push_switch(
-                    |(h, client): (TransportHeader, ClientInfo)| match h {
-                        TransportHeader {
-                            port,
-                            name: None,
-                            protocol: None,
-                        } => Ok(svc::Either::A(port)),
-                        TransportHeader {
-                            port,
-                            name: Some(name),
-                            protocol,
-                        } => Ok(svc::Either::B(GatewayTransportHeader {
-                            target: NameAddr::from((name, port)),
-                            protocol,
-                            client,
-                        })),
-                        TransportHeader {
-                            name: None,
-                            protocol: Some(_),
-                            ..
-                        } => Err(RefusedNoTarget),
+                    {
+                        let policies = policies.clone();
+                        move |(h, client): (TransportHeader, ClientInfo)| -> Result<_> {
+                            match h {
+                                TransportHeader {
+                                    port,
+                                    name: None,
+                                    protocol: None,
+                                } => {
+                                    let allow = policies.check_policy(OrigDstAddr(
+                                        (client.local_addr.ip(), port).into(),
+                                    ))?;
+                                    let tls = tls::ConditionalServerTls::Some(
+                                        tls::ServerTls::Established {
+                                            client_id: Some(client.client_id),
+                                            negotiated_protocol: client.alpn,
+                                        },
+                                    );
+                                    let _permit =
+                                        allow.check_authorized(client.client_addr, tls)?;
+                                    // TODO(ver) Use the permit's labels in metrics...
+                                    Ok(svc::Either::A(port))
+                                }
+                                TransportHeader {
+                                    port,
+                                    name: Some(name),
+                                    protocol,
+                                } => {
+                                    let allow = policies.check_policy(client.local_addr)?;
```

Comment:
why does the case where the transport header includes a name and a port check the client's local addr, while the case where there's no name check the client's local addr IP and the port from the transport header?

is this an oversight, or is it on purpose? and if it's on purpose, can we have a comment here explaining why it's different?

Reply:
when it includes a name/port, it's a gateway configuration targeting some other service; so the policy check determines whether the client can contact the _gateway_ itself. will add a comment